Runpod data fetcher (continued from PR#5930)

adocherty opened the pull request:
<!-- Describe the changes in this PR -->
## Context

This PR is continuing from https://github.com/skypilot-org/skypilot/pull/5930

As I don't have permissions for the original PR branch I've forked it and addressed the reviewer comments, plus fixed a few issues that I identified. Unfortunately, as I need to create a new PR from my branch I'm addressing the reviewer comments from the previous PR inline here:

## Reviewer comments from PR5930

I've addressed reviewer comments as follows :

[sky/catalog/data_fetchers/fetch_runpod.py](https://github.com/skypilot-org/skypilot/pull/5930/files/6bdb99f00755339a179857dce8067577f14ce98f#diff-4690f7151707b6de21ba86e2bf729acea2c207e7585a06847b72532733e6e0e0) Line 128

```python
    return round(price, 2)

def get_gpu_counts(max_count: int) -> List[int]:
```

> **[[andylizf](https://github.com/andylizf)]** is this function really used?

[adocherty](https://github.com/adocherty) response:
**Not that I can see. I’ve removed the get_gpu_counts function.**

[sky/catalog/data_fetchers/fetch_runpod.py](https://github.com/skypilot-org/skypilot/pull/5930/files/6bdb99f00755339a179857dce8067577f14ce98f#diff-4690f7151707b6de21ba86e2bf729acea2c207e7585a06847b72532733e6e0e0) Line 120

```yaml
    if 'errors' in result:
        raise ValueError(f'GraphQL errors: {result["errors"]}')

    return result['data']['gpuTypes'][0]
```

> **[[andylizf](https://github.com/andylizf)](https://github.com/andylizf)**  should have some assertions here?

[adocherty](https://github.com/adocherty) response:
 **I wrapped this access in a try/except loop to catch any errors more cleanly than multiple assertions**
 

[sky/catalog/data_fetchers/fetch_runpod.py](https://github.com/skypilot-org/skypilot/pull/5930/files/6bdb99f00755339a179857dce8067577f14ce98f#diff-4690f7151707b6de21ba86e2bf729acea2c207e7585a06847b72532733e6e0e0)  Line 150

```yaml
   return sorted(counts)

- def format_gpu_name(gpu_type: Dict) -> str:
+ def format_gpu_name(gpu_type: Dict[str, Any]) -> str:
```

[adocherty](https://github.com/adocherty) response:
**Applied this change**
 
Line 182
```yaml

    # Fall back to defaults if values are None
    # scale default value by gpu_count
    vcpus = DEFAULT_VCPUS * gpu_count if vcpus is None else vcpus
```

> **[[andylizf](https://github.com/andylizf)](https://github.com/andylizf)** lets have more assertions for this vcpus here? like is a positive 

[adocherty](https://github.com/adocherty) response:
 **Put in some assertions for vpcus and memory as positive and vpcus as integer.**
 
[sky/catalog/data_fetchers/fetch_runpod.py](https://github.com/skypilot-org/skypilot/pull/5930/files/6bdb99f00755339a179857dce8067577f14ce98f#diff-4690f7151707b6de21ba86e2bf729acea2c207e7585a06847b72532733e6e0e0) Line 254

```python
        # Generate instances from GPU types
        instances = []
        for gpu in gpus:
            # initial gpu details. later, request specific quantity details
```

> **[[andylizf](https://github.com/andylizf)](https://github.com/andylizf)** is it necessary? seems its quite redundant

[adocherty](https://github.com/adocherty) response:
**The comment? I've refactored this a little to make it tidier by doing the following:**

1. use a list comprehension to get the instances using the gpu_id

```python
        instances = [
            get_instance_configurations(gpu['id'])
            for gpu in gpus
        ]
```

2. Moved all calls to `get_gpu_details` inside `get_instance_configurations` and only call this once for `gpu_count=1` 

```python
def get_instance_configurations(gpu_id: str) -> List[Dict]:
    """Generate instance configurations for a GPU type."""
    instances = []
    detailed_gpu_1 = get_gpu_details(gpu_id, gpu_count=1)
    base_gpu_name = format_gpu_name(detailed_gpu_1)
    gpu_counts = detailed_gpu_1.get('lowestPrice', {}).get('availableGpuCounts', [])

    for gpu_count in gpu_counts:
        # Get detailed GPU info for this count
        if gpu_count == 1:
            detailed_gpu = detailed_gpu_1
        else:
            detailed_gpu = get_gpu_details(gpu_id, gpu_count)

```

[sky/catalog/data_fetchers/fetch_runpod.py](https://github.com/skypilot-org/skypilot/pull/5930/files/6bdb99f00755339a179857dce8067577f14ce98f#diff-4690f7151707b6de21ba86e2bf729acea2c207e7585a06847b72532733e6e0e0) Line 42

> **[kevinmingtarja](https://github.com/kevinmingtarja)** I think the position of Price and SpotPrice is reversed here, looking at https://github.com/skypilot-org/skypilot-catalog/blob/master/catalogs/v7/runpod/vms.csv
```
Suggested change
41 - 'Price',
42 - 'SpotPrice',
41 + 'SpotPrice',
42 + 'Price',
```

[adocherty](https://github.com/adocherty) response:
**I've checked this and you are correct, I've applied the changes.**

## Other changes

As the original author noted, there are both price differences and differences in the RAM quantities shown by this script and the original `vms.csv` here https://github.com/skypilot-org/skypilot-catalog/blob/master/catalogs/v7/runpod/vms.csv

This is a comparison of the top 5 rows of the new output from the script in this PR:

InstanceType | AcceleratorName | AcceleratorCount | vCPUs | MemoryGiB | GpuInfo | Region | SpotPrice | Price | AvailabilityZone
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
1x_A100-80GB_SECURE | A100-80GB | 1.0 | 8 | 117 | A100-80GB | CA | 0.82 | 1.64 | CA-MTL-1
1x_A100-80GB_SECURE | A100-80GB | 1.0 | 8 | 117 | A100-80GB | CA | 0.82 | 1.64 | CA-MTL-2
1x_A100-80GB_SECURE | A100-80GB | 1.0 | 8 | 117 | A100-80GB | CA | 0.82 | 1.64 | CA-MTL-3
1x_A100-80GB_SECURE | A100-80GB | 1.0 | 8 | 117 | A100-80GB | CZ | 0.82 | 1.64 | EU-CZ-1
1x_A100-80GB_SECURE | A100-80GB | 1.0 | 8 | 117 | A100-80GB | NL | 0.82 | 1.64 | EU-NL-1


With the current catalog:

InstanceType | AcceleratorName | AcceleratorCount | vCPUs | MemoryGiB | GpuInfo | Region | SpotPrice | Price | AvailabilityZone
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
1x_A100-80GB_SECURE | A100-80GB | 1.0 | 8.0 | 80.0 | A100-80GB | CA | 1.19 | 1.64 | CA-MTL-1
1x_A100-80GB_SECURE | A100-80GB | 1.0 | 8.0 | 80.0 | A100-80GB | CA | 1.19 | 1.64 | CA-MTL-2
1x_A100-80GB_SECURE | A100-80GB | 1.0 | 8.0 | 80.0 | A100-80GB | CA | 1.19 | 1.64 | CA-MTL-3
1x_A100-80GB_SECURE | A100-80GB | 1.0 | 8.0 | 80.0 | A100-80GB | CZ | 1.19 | 1.64 | EU-CZ-1
1x_A100-80GB_SECURE | A100-80GB | 1.0 | 8.0 | 80.0 | A100-80GB | IS | 1.19 | 1.64 | EUR-IS-1


1. Memory differences

**The script in the original PR used the CPU RAM memory not the GPU memory which is in the current catalog. I've modified this script to return the GPU memory to replicate the current catalog.**

2. Price differences

There are catalog price differences particularly with the spot prices. This is most likely due to price changes by RunPod and the costs returned by the script match the spot prices manually checked on the RunPod website:
<img width="271" height="153" alt="image" src="https://github.com/user-attachments/assets/e0ed0432-bf55-479d-8fc3-53d318526fc4" />

3. There are differences in the ordering of the availability zones, as seen in the last line of the table.

The script in this PR orders by availability zone, rather than the region. I don't see this as being a big issue but I can fix it up if required?

```
df.sort_values(['AcceleratorName', 'InstanceType', 'AvailabilityZone'], inplace=True)
```

## Not yet addressed

I haven't yet modified the code to get CPU instances into the RunPod catalog. I'm happy to include them either in this PR or another but I'd like to ensure I'm on the right path first.

> Hi @pirtleshell, thanks for writing this PR! FYI We're also planning on adding CPU instances to the Runpod catalog: https://github.com/skypilot-org/skypilot-catalog/pull/140
> I generated the CPU instances list by modifying the script from this PR. Would you mind adding it to this PR as well? Or we could do it in a follow-up PR too. I can share the diffs I made to your script to support fetching CPU instances.
> In any case, we should do that before merging https://github.com/skypilot-org/skypilot-catalog/pull/133, as to not override the CPU instances we will be adding.

See original comment here:  [kevinmingtarja](https://github.com/kevinmingtarja) commented (https://github.com/skypilot-org/skypilot/pull/5930#issuecomment-3138266562)


## Testing

<!-- Describe the tests ran -->

For testing I've followed the original PR and generated the new catalog `vms.cvs` file using the following manual command:
```
RUNPOD_API_KEY=read-only-api-key python sky/catalog/data_fetchers/fetch_runpod.py --output-dir temp
RunPod Service Catalog saved to temp/vms.csv
```

<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: install pre-commit (auto-check on commit) or `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `/smoke-test` (CI) or `pytest tests/test_smoke.py` (local)
- [ ] Relevant individual tests: `/smoke-test -k test_name` (CI) or `pytest tests/test_smoke.py::test_name` (local)
- [ ] Backward compatibility: `/quicktest-core` (CI) or `pytest tests/smoke_tests/test_backward_compat.py` (local)

<!-- CI commands (/-prefixed) can only be triggered by repo members -->

adocherty commented:
Thanks for the review @kevinmingtarja and @Michaelvll!

I've addressed your review comments as above, let me know what you think.

I’ve also pulled the calculation of `gpu_info` `spot_price` and `base_price` outside the region loop as they are all pulled from `detailed_gpu` which doesn’t change with region.

adocherty commented:
After further testing I’ve noticed that the number of VCPUs and available counts returned by RunPod API in the `lowestPrice` data **changes dynamically over time well under the 7 hour refresh time.** This doesn’t seem to be in the 
 spirit of the RunPod catalog. 

I think that we should aim to list the configurations that are possible on the cloud rather than capturing what is currently available.  SkyPilot then handles the query with the API to determine if the requested configuration is currently available.

Therefore, In order to not have a catalog that changes every time this script is run I’ve taken a different approach:

As guided by @kevinmingtarja, I’ve made a hard-coded mapping of GPU to both the number of VCPUs and the max count of GPUs available.  This was done as follows:

1. If there exists an entry in the current v8 catalog, take this as the vCPU count
2. If not, and there’s a listing on the pricing page, take this as the vCPU count
3. Otherwise, take the minimum of the vCPUs returned by the RunPod API over a 2 hour polling

This provides the following:

- The current catalog is replicated in the GPUs that are overlapping
- New GPUs are supported

The comparison between VCPUs for each GPU from the current catalog, API, and pricing page are listed below:

|  | vCPU current catalog (v8) | min vCPUs API | max vCPUs API | Pricing Page |
| --- | --- | --- | --- | --- |
| A100-80GB | **8** | 8 | 31 | 8 |
| A100-80GB-SXM | **16** | 16 | 32 | 16 |
| A30 |  | **8** | 9 |  |
| A40 | **9** | 9 | 9 | 9 |
| B200 | **36** | 24 | 28 | 28 |
| H100 | **16** | 16 | 31 | 16 |
| H100-NVL | **16** | 16 | 19 | 16 |
| H100-SXM | **20** | 20 | 28 | 16 |
| H200-SXM | **20** | 12 | 24 | 24 |
| L4 | **4** | 8 | 8 | 12 |
| L40 | **16** | 9 | 9 | 8 |
| **L40S** | **12** | **16** | **16** | **16** |
| MI300X | **24** | 24 | 24 |  |
| RTX2000-Ada |  | **6** | 6 |  |
| RTX3070 |  | **8** | 16 |  |
| RTX3080 |  | **8** | 8 |  |
| RTX3080-Ti |  | **8** | 25 |  |
| RTX3090 | **16** | 4 | 32 | 16 |
| RTX3090-Ti |  | **12** | 14 |  |
| RTX4000-Ada | **9** | 8 | 16 |  |
| RTX4090 | **16** | 5 | 32 | 6 |
| RTX5000-Ada |  | **6** | 14 |  |
| RTX5080 |  | **5** | 5 |  |
| RTX5090 |  | 5 | 16 | **9** |
| RTX6000-Ada | **14** | 10 | 24 |  |
| RTXA4000 | **6** | 4 | 14 |  |
| RTXA4500 | **12** | 7 | 12 |  |
| RTXA5000 | **6** | 8 | 9 | 9 |
| RTXA6000 | **4** | 8 | 8 | 9 |
| RTXPRO6000 |  | **14** | 14 |  |
| RTXPRO6000-MaxQ |  | **18** | 18 |  |
| RTXPRO6000-WK |  | **32** | 32 |  |
| V100-SXM2-32GB |  | **20** | 20 |  |

I note that in the current config file all GPUs have VCPUs scaled by the number of GPUs except L40S:

| InstanceType | AcceleratorName | AcceleratorCount | vCPUs | MemoryGiB | GpuInfo | Region | SpotPrice | Price | VRAM_pu | vCPU_pu |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 1x_L40S_SECURE | L40S | 1 | 12 | 62 | L40S | CA | 0.86 | 0.43 | 62 | 12 |
| 2x_L40S_SECURE | L40S | 2 | 64 | 376 | L40S | CA | 1.72 | 0.86 | 188 | 32 |
| 4x_L40S_SECURE | L40S | 4 | 256 | 1504 | L40S | CA | 3.44 | 1.72 | 376 | 64 |
| 8x_L40S_SECURE | L40S | 8 | 1024 | 6016 | L40S | CA | 6.88 | 3.44 | 752 | 128 |

It looks like there is an error for L40S in the current catalog, and I go with VCPUs=16 which matches both the API and the pricing page information.

For the maximum count of GPUs I take this as 8 for all GPUs except A40 which I take to be 10.  This comes again from monitoring the API for a couple of hours.  There are some GPUs that do not exceed 4, I can optionally reduce the count of these if required.

To summarise, the changes over current catalog are now:

- L40S has been amended to have VCPUs=16 per GPU
- Number of VCPUs for new GPUs is included
- Pods with GPU counts from 1 to 8 are supported, and these are adjustable by GPU (original catalog only includes 1,2,4 and 8 for all GPUs)

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+    return round(price, 2)
+
+
+def format_gpu_name(gpu_type: Dict[str, Any]) -> str:
```
It may be helpful to centralize this helper somewhere. I don't know how the rest of the repo works but this feels like something you may want to centralize

kevinmingtarja commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -194,10 +338,15 @@ def get_instance_configurations(gpu_id: str) -> List[Dict]:
-    for gpu_count in gpu_counts:
+    # If the GPU isn't in DEFAULT_GPU_INFO we default to a max of 8 GPUs
+    if base_gpu_name in DEFAULT_GPU_INFO:
+        max_gpu_count = DEFAULT_GPU_INFO['base_gpu_name'].get(
```
Small typo: `'base_gpu_name' -> base_gpu_name`

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+DEFAULT_MAX_GPUS = 8
+DEFAULT_GPU_INFO: Dict[str, Dict[str, Union[int, float]]] = {
+    'A100-80GB': {
+        'vcpus': 8.0,
```
should be 12, but generally points to something that would be difficult to maintain. These numbers are hard-coded but if I were to implement this I'd probably try to reduce hard-coding as much as possible. It's likely better to decide on an imperfect source of truth (the api or parsing the html pricing page), and leave it at that.

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+        'max_count': 10
+    },
+    'B200': {
+        'vcpus': 36.0,
```
should be 24, but anyway I think I won't review each one individually, but it should point to the fact this approach encodes inevitable drift.

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+        'vcpus': 18.0,
+        'max_count': 8
+    },
+    'RTXPRO6000-WK': {
```
why do you do RTXPRO6000 instead of RTX-PRO-6000, the later name updates would suggest that spaces are replaced with - but in this it appears to be using a different heuristic. Perhaps it would be a good idea to centralize and standardize how the names of GPUs are defined. Then allow for each fetcher script to define its own mapping from the centralized GPU name list and the gpu name list for that particular vendor.

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+
+# Mapping of regions to their availability zones
+REGION_ZONES = {
+    'CA': ['CA-MTL-1', 'CA-MTL-2', 'CA-MTL-3'],
```
these will get out of date, I suspect you can source this information dynamically via the api or html parsing. Either way, it's going to get out of sync and need to be updated so an imperfect source may be better.

adocherty replied:
Agrees, I don't like hard coding these; however, I don't know of any way to get this information dynamically. Do you know how we can do that?

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+
+def format_price(price: float) -> float:
+    """Format price to two decimal places."""
+    return round(price, 2)
```
it may be helpful to centralize pricing information formatting, this will likely become quite difficult to maintain. Furthermore, using floats for prices is an anti-pattern I would advise against. If the runpod API uses floats then it's best to centralize a way to convert this to an integer or use a package to manage it. But I'd also generally advise skypilot to stay away from anything related to accounting. Because it will add a tremendous maintenance burden and is not directly aligned with the core ethos: deploy and manage workloads. The vendors integrate their own accounting. If accounting is a highly requested feature, it's better for later stages when you can put more care and time into it, and integrating accounting information early on will likely result in bugs and incorrect information which will erode trust with your userbase unnecessarily. that's just my two cents

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+    base_name = (
+        gpu_type['displayName']
+        # handle formatting names of RTX GPUs
+        .replace('RTX PRO ', 'RTXPRO')
```
perhaps you can ignore my previous comment about RTX Pro

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+        # Returns vcpus already scaled by the GPU count
+        vcpus = gpu_type.get('lowestPrice', {}).get('minVcpu')
+    else:
+        vcpus = vcpus * gpu_count
```
This is making an assumption. In my opinion it's better to report to the user that the number of cpus available is unknown rather than making an assumption and being wrong.

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+
+    # Convert the counts, vCPUs, and memory to float
+    # for consistency with skypilot's catalog format
+    return {
```
NOTE: you can use typed dicts from the typing library to enforce standard keys and value types, which is likely going to be helpful in a case liek this where you're passing around dicts instead of objects. I'd advise looking more into it

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+    else:
+        vcpus = vcpus * gpu_count
+
+    # This is the VRAM memory per GPU, not the CPU memory
```
I advise modifying the variable names instead of having a comment, unless "memory" is well-defined in skypilot's domain specific language as being gpu memory. But having to specificy this each time you index memory would be difficult and prone to error

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+
+    # If the GPU isn't in DEFAULT_GPU_INFO we default to a max of 8 GPUs
+    if base_gpu_name in DEFAULT_GPU_INFO:
+        max_gpu_count = DEFAULT_GPU_INFO['base_gpu_name'].get(
```
is this supposed to use base_gpu_name instead of 'base_gpu_name'?

alita-moore commented on sky/catalog/runpod_catalog.py:
```diff
@@ -12,7 +12,11 @@
+# Runpod has no set updated schedule for their catalog. We pull the catalog
+# every 7 hours to make sure we have the latest information.
+_PULL_FREQUENCY_HOURS = 7
+_df = common.read_catalog('runpod/vms.csv',
```
are you actually rerunning the script every 7 hours though or just pulling from the csv? It may be helpful to review if it's actually running the job again because otherwise it wouldn't be updating.

kevinmingtarja replied:
Once this script is merged, we will add a new github workflow in our other repo https://github.com/skypilot-org/skypilot-catalog, and there we will run the script every 7h.

alita-moore commented on sky/catalog/data_fetchers/fetch_runpod.py:
```diff
@@ -0,0 +1,466 @@
+    parser = argparse.ArgumentParser(
+        description='Update RunPod catalog for SkyPilot')
+    parser.add_argument('--output-dir',
+                        default='runpod',
```
this may be helpful to set dynamically, wouldn't this assume that the script is being run with the default path being `~/.sky/catalogs/v8` or whatever?

alita-moore reviewed (commented):
I have a few philosophical suggestions, and there's one possible typo with the use of base_gpu_name. I think if you disagree with my comments about tech debt or feel this is a one-off script then I'd say it's good to go. But otherwise I think there are opportunities for abstractions, which may be helpful given that this fetching logic is central to the core functionality of skypilot and getting this script correct coudl save you many hours down the line when the script inevitably becomes obsolete / outdated.

adocherty commented:
<b>Updates</b>

I've updated the PR and moved to hard-code the CPU memory and the vCPU count as extracted from the API by polling it over a 4 hour period and taking the minimum of both.

Note that **this changes the catalog from the current version **. In particular, the current catalog has the GPU memory not the CPU memory (see the table below).

<p>The table below shows the vCPUS, Memory and GPU memory obtained from calls to the API compared to the same information from the pricing page and the current catalog (v8).</p>
<p>In particular, notice that **the current catalog lists the GPU memory not the CPU memory in the catalog**, if we move to use the CPU memory we will change the catalogue.</p>
<p>I was trying to add functionality with minimal changes to the current catalog, so I’m worried about the potential side effects of this change on current code using the functionality.</p>
<p>I have nonetheless decided to move to using the results of the API calls and in the future we can hopefully transition to make this automated rather than hard-coding it.</p>

AccelleratorName | API Calls |   |   |   | Pricing Page |   | Current Catalog | (v8)
-- | -- | -- | -- | -- | -- | -- | -- | --
  | vCPUs | MemoryInGIB | GPU Memory | Max gpuCount | vCPUs | GPU Memory | vCPUs | MemoryInGIB
A100-80GB | 8 | 117 | 80 | 8 | 8 | 80 | 8 | 80
A100-80GB-SXM | 16 | 117 | 80 | 8 | 16 | 80 | 16 | 80
A30 | 12 | 39 | 24 | 8 |   |   |   |  
A40 | 9 | 48 | 48 | 10 | 9 | 48 | 9 | 48
B200 | 28 | 180 | 180 | 8 | 28 | 180 | 36 | 180
H100 | 16 | 176 | 80 | 8 | 16 | 80 | 16 | 80
H100-NVL | 16 | 94 | 94 | 10 | 16 | 94 | 16 | 94
H100-SXM | 20 | 125 | 80 | 8 | 20 | 80 | 20 | 80
H200-SXM | 12 | 188 | 141 | 8 | 24 | 141 | 20 | 141
L4 | 8 | 45 | 24 | 10 | 12 | 24 | 4 | 24
L40 | 9 | 125 | 48 | 10 | 8 | 48 | 16 | 48
L40S | 12 | 62 | 48 | 8 | 16 | 48 |   |  
MI300X | 24 | 283 | 192 | 8 |   |   | 24 | 192
RTX2000-Ada | 6 | 31 | 16 | 8 |   |   |   |  
RTX3070 | 8 | 30 | 8 | 8 |   |   |   |  
RTX3080 | 8 | 14 | 10 | 4 |   |   |   |  
RTX3080-Ti | 8 | 18 | 12 | 5 |   |   |   |  
RTX3090 | 4 | 25 | 24 | 8 | 16 | 24 | 16 | 24
RTX3090-Ti | 8 | 24 | 24 | 9 |   |   |   |  
RTX4000-Ada | 8 | 47 | 20 | 8 |   |   | 9 | 20
RTX4080 | 8 | 22 | 16 | 5 |   |   |   |  
RTX4080-SUPER | 12 | 62 | 16 | 6 |   |   |   |  
RTX4090 | 5 | 29 | 24 | 8 | 6 | 24 | 16 | 24
RTX5000-Ada | 6 | 62 | 32 | 8 |   |   |   |  
RTX5080 | 5 | 30 | 16 | 8 |   |   |   |  
RTX5090 | 6 | 46 | 32 | 8 | 9 | 32 |   |  
RTX6000-Ada | 10 | 62 | 48 | 8 | 10 | 48 | 14 | 48
RTXA4000 | 6 | 35 | 16 | 12 |   |   | 6 | 16
RTXA4500 | 7 | 30 | 20 | 4 |   |   | 12 | 20
RTXA5000 | 3 | 25 | 24 | 10 | 9 | 24 | 6 | 24
RTXA6000 | 8 | 50 | 48 | 10 | 9 | 48 | 4 | 48
RTXPRO6000 | 14 | 125 | 96 | 9 |   |   |   |  
RTXPRO6000-MaxQ | 18 | 215 | 96 | 7 |   |   |   |  
RTXPRO6000-WK | 12 | 186 | 96 | 4 |   |   |   |  
V100-SXM2 | 10 | 62 | 16 | 8 |   |   |   |  
V100-SXM2-32GB | 20 | 93 | 32 | 4 |   |   |   |  

**Summary**

<p>I’ve updated the code to address the bug found, transition to use CPU memory in the catalog not GPU memory and additionally added a GpuInfo column of JSON data modelled from the fetch_nebius.py script.</p>
<p>This PR will allow us to move forward and gives the following:</p>
<ul>
<li>An automated generation of RunPod catalog</li>
<li>Price changes will be kept up to date</li>
<li>Will catch new GPUs being added automatically</li>
</ul>
<p>However, there are some issues to solve for moving forward:</p>
<ul>
<li>How can we bridge the divide between RunPods dynamic catalog (pod configuations change in minutes) and SkyPilot’s static catalog (refreshed every 7 hours)?</li>
<li>How can we move away from a hard-coded list of vCPU and memory configurations?</li>
</ul>
<!-- notionvc: 75fe653a-d330-4398-b7ae-4e3452d4b74e -->

kevinmingtarja commented:
Superseded by #7460. Thanks again @adocherty !
